File: merlin-server/constraints-dsl-compiler/src/libs/segment.ts
import {Interval} from "./interval";
import type {Boundable} from "./timeline";

export class Segment<V> implements Boundable<Segment<V>> {
  public readonly interval: Interval;
  public readonly value: V;

  constructor(value: V, interval: Interval) {
    this.interval = interval;
    this.value = value;
  }

  public transpose(): Segment<NonNullable<V>> | undefined {
    if (this.value === undefined || this.value === null) return undefined;
    else return new Segment<NonNullable<V>>(this.value, this.interval);
  }

  public mapValue<W>(f: (v: V, i: Interval) => W): Segment<W> {
    return new Segment<W>(f(this.value, this.interval), this.interval);
  }

  public mapInterval(f: (v: V, i: Interval) => Interval): Segment<V> {
    return new Segment<V>(this.value, f(this.value, this.interval));
  }

  bound(bounds: Interval): Segment<V> | undefined {
    const intersection = Interval.intersect(bounds, this.interval);
    if (intersection.isEmpty()) return undefined;
    else return new Segment(this.value, intersection);
  }


}

File: timeline/src/main/kotlin/gov/nasa/jpl/aerie/timeline/payloads/Segment.kt
package gov.nasa.jpl.aerie.timeline.payloads

import gov.nasa.jpl.aerie.timeline.Interval
import gov.nasa.jpl.aerie.timeline.util.coalesceList

/**
 * A generic container that associates a value with an interval on a timeline.
 */
data class Segment<V>(/***/ override val interval: Interval, /***/ val value: V): IntervalLike<Segment<V>> {
  /**
   * Create a new segment on the same interval, with a new value derived from this segment.
   *
   * In most cases using [withNewValue] will result in simpler-looking code.
   *
   * @param f a function that takes `this` as argument and produces a new value.
   */
  fun <W> mapValue(f: (Segment<V>) -> W) = Segment(interval, f(this))

  /**
   * Create a new segment with the same value, on a new interval derived from this segment.
   *
   * In most cases using [withNewInterval] will result in simpler-looking code.
   *
   * @param f a function that takes `this` as argument and produces a new interval.
   */
  fun mapInterval(f: (Segment<V>) -> Interval) = Segment(f(this), value)

  /**
   * Creates a new segment with the same value on a new interval.
   *
   * @param i the new interval
   */
  override fun withNewInterval(i: Interval) = Segment(i, value)

  /**
   * Creates a new segment on the same interval with a new value.
   *
   * @param w the new value
   */
  fun <W> withNewValue(w: W): Segment<W> = Segment(interval, w)

  /**
   * Checks whether this segment's value equals another, using the basic equality operator.
   *
   * Used mostly for the [coalesceList] operation.
   */
  fun valueEquals(other: Segment<V>) = value == other.value

  /***/
  companion object {
    /**
     * Delegates to the constructor.
     *
     * Can be more convenient in Java by avoiding the `new` keyword.
     */
    @JvmStatic fun <V> of(interval: Interval, value: V) = Segment(interval, value)
  }
}

/**
 * Converts a non-null segment of a maybe-null value into a maybe-null segment of a non-null value.
 */
fun <V> Segment<V?>.transpose() = if (value == null) null else Segment(interval, value)
